### ISSUE TITLE
Remove `StateTransition` schedule in favor of using observers

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#15133]
## What problem does this solve or what need does it fill?

Updating the app state is a very flexible and expressive operation, which is fundamentally deferred. States typically changes only once per frame, after `PostUpdate`.

This can lead to unwanted delays when handling multiple dependent transitions at once.

## What solution would you like?

As discussed in #15127, it would be nice to remove this meta-schedule, and instead run the various `OnEnter` / `OnExit` / `OnTransition` schedules on demand, leveraging the new observers. This is a good fit for "highly complex rarely exercised logic", and resolves the delay problem mentioned above while simplifying the mental model for consumers.

#15127 should be tackled in concert with this, removing `NextState` completely, and relying entirely on commands to handle transitions.

## What alternative(s) have you considered?

Users can manually add more state transition systems to their schedule, including in their own schedules that run inside of `Main`. This is somewhat messy, nonstandard and hard to discover.

## Additional context

Discussed briefly on Discord with @MiniaczQ [here](https://discordapp.com/channels/691052431525675048/749335865876021248/1282862231570092033).

This design was not originally taken (including in the substates refactor) because observers didn't exist at the time!

**Comment by benfrankel** [comment]
If this is meant to replace the fixed once-per-frame state transitions schedule instead of providing an additional option, I'm strongly opposed. If not, I'd also like to see an example use case that isn't already covered by `world.run_schedule` or a custom command that calls that (why don't we have a `commands.run_schedule`?).

**Comment by alice-i-cecile** [comment]
Can you explain why you prefer a fixed once-per-frame state transition?

And yeah, we can just add commands.run_schedule.

**Comment by MiniaczQ** [comment]
I'm actually not super sold on the importance of "run state transition at any point in time", I think we should have it as a niche feature and keep synchronized transitions as the main feature.

The biggest use for state transitions is spawning in entities, which may be processed in multiple schedules.
If entities are spawned at arbitrary times within the schedule, it's hard to guarantee all of them are processed correctly every time.

**Comment by benfrankel** [comment]
> for example if you have a Level(usize) state, and you generate / load a new level on enter
you don't want arbitrarily many Level transitions in the same frame, at arbitrary points in the schedule 
probably
there may be some use cases where you do want that that i'm not aware of

_From Pyrious on Discord_

> or if you're in the middle of the frame and trigger a state change, then the last half of the frame you're running systems in the new state
which likely isn't intended

_From doot on Discord_

**Comment by benfrankel** [comment]
> I'd like to see an example use case that isn't already covered by `world.run_schedule` or a custom command that calls that (why don't we have a `commands.run_schedule`?).

Actually, `world.run_schedule` alone requires you to run _all_ state types' transition hooks via the `StateTransitions` schedule. You don't get the flexibility of only running the transition hooks of _one_ state type and its descendant (computed / sub) states' transition hooks.

There is actually a proposal in the hackmd called "Flush hooks as observers" where the `StateTransitions` schedule remains, but instead of running a bunch of regular systems with ordering, it triggers an event for each state type, and observers will respond to the event by running state transition hooks. This enables triggering the event for a particular state type outside of the `StateTransitions` schedule, wherever you want, while keeping the once-per-frame flush point intact.

But again, whether this proposal should be implemented will depend on the existence of an example use case, which I don't have but I also haven't thought much about.

**Comment by Shatur** [comment]
I think it's a great idea.

In [bevy_replicon](https://github.com/projectharmonia/bevy_replicon), we have separate states for the client and server, but we can't use Bevy states. All messaging backends update the state in `PreUpdate`, and we also need to react to it in `PreUpdate` because Replicon is also a library. However, the state updates after `StateTransition`, which means `run_if(in_state(..))` would result in a one-frame delay. We could add our custom schedule after `StateTransitions`, but it's weird and reduces parallelism. This is why our states aren't Bevy states 😢 

With the proposed idea, we would be able to perform the transition inside `PreUpdate` and avoid the one-frame delay.

**Comment by benfrankel** [comment]
I mean I think `StateTransition` schedule should run before `PreUpdate` anyways. I recently made this change in `pyri_state`.

**Comment by Shatur** [comment]
> I mean I think the StateTransition schedule should run before PreUpdate anyways

This would result in one frame delay between messaging backends and Replicon.

> Or possibly support not inserting the schedule at all so power users can insert it themselves wherever they prefer.

The ability to put something like `StateTransition<ClientState>` and `StateTransition<ServerState>` anywhere would still result in one frame delay 😢 For my use case I need to have a state transition inside `PreUpdate`.

This is why I so excited about using observers for state transitions.

**Comment by benfrankel** [comment]
In `pyri_state` I provide a [system set](https://docs.rs/pyri_state/latest/pyri_state/schedule/resolve_state/enum.ResolveStateSystems.html#variant.Compute) to optionally compute the next value for a state within the `StateTransition` schedule. In this design, the messaging backends would be using that instead of setting the next state in a normal schedule (which has a 1-frame delay by design). And then `PreUpdate` would run after the state transition.

**Comment by Shatur** [comment]
Answered in Discord, but I'll duplicate it here just in case 🙂 

> All messaging backends update their state and receive messages in `PreUpdate`.
After that they need to set the state for Replicon and pass all messages to it.
And finally replicon reacts on messages.
I can't ask messaging backends to put message receiving before `PreUpdate` because they need to run some logic in `PreUpdate` first.

**Comment by Shatur** [comment]
I found a nice way to solve my problem: simply run the systems additionally in `OnEnter`.  
This works because my systems run in `PreUpdate` if specific state is set. Backends also change the state in `PreUpdate`  
So when the state changes, the systems still run in that frame - just a bit later, during `OnEnter`.

Here's my PR for reference: https://github.com/simgine/bevy_replicon/pull/565

As for the proposal, after giving it some thought, I'm not sure having multiple state transitions per frame is a good idea.
So I think I agree with @benfrankel here. Maybe we could just have a way to customize where the transition happens?
There’s also a more recent issue taking a slightly different approach: https://github.com/bevyengine/bevy/issues/18583, though I'm not entirely sold on it either…
